Add noinlineerr linter

AlwxSin opened the pull request:
Adding [noinlineerr](https://github.com/AlwxSin/noinlineerr) linter.

Purpose - allow only one style of error handling.
<!--

WARNING:

We use Dependabot to update dependencies (linters included).
The updates happen at least automatically once a week (Sunday 11am UTC).

No pull requests to update a linter will be accepted unless you are the author of the linter AND specific changes are required.

-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/golangci/golangci-lint?pullRequest=5826) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/golangci/golangci-lint?pullRequest=5826) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/golangci/golangci-lint?pullRequest=5826) it.</sub>

ldez commented:
In order for a pull request adding a linter to be reviewed, the linter and the PR must follow some requirements.

- [ ] The CLA must be signed

### Pull Request Description

- [ ] It must have a link to the linter repository.
- [ ] It must provide a short description of the linter.

### Linter

- [ ] It must not be a duplicate of another linter or a rule of a linter (the team will help to verify that).
- [ ] It must have a valid license (AGPL is not allowed), and the file must contain the required information by the license, ex: author, year, etc.
- [ ] It must use Go version <= 1.23.0
- [ ] The linter repository must have a CI and tests.
- [ ] It must use [`go/analysis`](https://golangci-lint.run/contributing/new-linters/).
- [ ] It must have a valid tag, ex: `v1.0.0`, `v0.1.0`.
- [ ] It must not contain `init()`.
- [ ] It must not contain `panic()`.
- [ ] It must not contain `log.Fatal()`, `os.Exit()`, or similar.
- [ ] It must not modify the AST.
- [ ] It must not have false positives/negatives (the team will help to verify that).
- [ ] It must have tests inside golangci-lint.

### The Linter Tests Inside Golangci-lint

- [ ] They must have at least one std lib import.
- [ ] They must have integration tests without configuration (default).
- [ ] They must have integration tests with configuration (if the linter has a configuration).

### `.golangci.next.reference.yml`

- [ ] The file `.golangci.next.reference.yml` must be updated.
- [ ] The file `.golangci.reference.yml` must NOT be edited.
- [ ] The linter must be added to the lists of available linters (alphabetical case-insensitive order).
    - `enable` and `disable` options
- [ ] If the linter has a configuration, the exhaustive configuration of the linter must be added (alphabetical case-insensitive order)
    - The values must be different from the default ones.
    - The default values must be defined in a comment.
    - The option must have a short description.

### Others Requirements

- [ ] The files (tests and linter) inside golangci-lint must have the same name as the linter.
- [ ] The `.golangci.yml` of golangci-lint itself must not be edited and the linter must not be added to this file.
- [ ] The linters must be sorted in the alphabetical order (case-insensitive) in the `lintersdb/builder_linter.go` and `.golangci.next.reference.yml`.
- [ ] The load mode (`WithLoadMode(...)`):
    - if the linter uses `goanalysis.LoadModeSyntax` -> no `WithLoadForGoAnalysis()` in `lintersdb/builder_linter.go`
    - if the linter uses `goanalysis.LoadModeTypesInfo`, it requires `WithLoadForGoAnalysis()` in `lintersdb/builder_linter.go`
- [ ] The version in `WithSince(...)` must be the next minor version (`v2.X.0`) of golangci-lint.
- [ ] `WithURL()` must contain the URL of the repository.

### Recommendations

- [ ] The file `jsonschema/golangci.next.jsonschema.json` should be updated.
- [ ] The file `jsonschema/golangci.jsonschema.json` must NOT be edited.
- [ ] The linter repository should have a readme and linting.
- [ ] The linter should be published as a binary (useful for diagnosing bug origins).
- [ ] The linter repository should have a `.gitignore` (IDE files, binaries, OS files, etc. should not be committed)
- [ ] A tag should never be recreated.
- [ ] use `main` as the default branch name.

---

The golangci-lint team will edit this comment to check the boxes before and during the review.

The code review will start after the completion of those checkboxes (except for the specific items that the team will help to verify).

The reviews should be addressed as commits (no squash).

If the author of the PR is a member of the golangci-lint team, he should not edit this message.

**This checklist does not imply that we will accept the linter.**

AlwxSin commented:
@ldez thank you for checklist.

Fixed:
* It must use Go version <= 1.23.0
* They must have at least one std lib import.
* The version in WithSince(...) must be the next minor version (v2.X.0) of golangci-lint.
* added `.gitignore`

Quick question:
> The files (tests and linter) inside golangci-lint must have the same name as the linter.

I thought they already do – did I miss something? Could you clarify what exactly should be renamed or matched?

ldez commented:
I detected at least one false-positive:
```go

func _() {
	if ok, _ := strconv.ParseBool("1"); ok {
		fmt.Println("ok")
	}
}
```

The linter is very limited: it only handles explicit `error` type references, so false negatives can happen.

AlwxSin commented:
@ldez Thanks for the feedback!

Do you have any specific ideas in mind?

Some thoughts I had:

Adding an option to choose the preferred style (though I’m not sure if it makes sense to only prefer the inline variant).

Adding autofix support — though that’s a bit more complex for me at the moment and would need some digging in.

Happy to hear any other suggestions you might have!

ldez commented:
The proposed options are not what I expected.

- `allow-inline-in-tests` will not be used by golangci-lint because this must be handled inside the golangci-lint exclusions and not a the linter level.
- `allow-multi-assign` I don't see why this should be a specific topic.


I'm talking about a broader perspective of inlined `if`, not to the restricted scope of errors.

AlwxSin commented:
About the broader perspective — I get your point, but I’d prefer to keep the scope limited to `err := ...` cases for now.
This is where most readability issues tend to show up in practice, and it matches the linter's intent (and name — `noinlineerr`).

I’ve often seen inline error handling being overused or misused, whereas inline Init with other values is way less common —
so I focused on the pattern that was actually causing pain in the codebases I work with.

That said, building a separate linter to cover all inline `Init` statements (`if x := ...; cond {}`) could definitely be interesting —
but that feels like a broader concern and a different project altogether.

ldez commented:
ok, so we agree that if a new linter is added that handles the whole topic of inlined `if`, your linter will be deprecated inside golangci-lint and removed.

If we agree on that, we can accept this linter.

ldez reviewed (approved):
LGTM

G-Rath commented:
@ldez fyi it looks like this linter is not listed in https://golangci-lint.run/usage/linters/

ldez commented:
@G-Rath refresh the page: https://golangci-lint.run/usage/linters/#noinlineerr

bombsimon commented:
The linter is not enabled by default, no new linters are. If you don't agree with the style (I don't), don't enable the linter.

ldez commented:
This linter, like other linters, is related to style.
The style is personal, so I don't want to provide any recommendations about using or not using a linter.

ruanxin commented:
Hi @bombsimon @ldez , thanks for the quick feedback.

First, I have to apologise, after checking our golangci.yaml, I realised the issue happened because we have `default: all` enabled and then selectively disable specific linters. That means when we upgraded, `noinlineerr` became active automatically.

But this actually highlights a point: our team (and I suspect many others) **treat golangci-lint as a trusted guideline for recommended code style.** When new linters are added, the implicit expectation is that they are **widely recognized, well-accepted, and non-controversial**. In this case, as I mentioned earlier, many in the Go community consider banning inline error handling a code smell rather than a style preference, as I mentioned in the [Uber’s guideline](https://github.com/uber-go/guide/blob/master/style.md#reduce-scope-of-variables), which explicitly encourages the opposite in certain cases.

By including a specific linter, golangci-lint still implicitly signals that this style is worth enforcing, since golangci-lint is trusted by a huge portion of the Go community, that signal matters I believe.

However, if the position of golangci-lint is that style linters are **purely neutral and not meant to guide or educate on best practices**, then I understand, but that’s a different philosophy than what I think many users expect. Perhaps we’ve simply placed too much expectation on golangci-lint as a de facto style authority, but that expectation is real.

Follow-up question: if someone later proposed a linter that enforces the use of inline error handling, which directly conflicts with this `noinlineerr`,  as it's exactly the opposite style, but is backed by reputable guidelines. Would you accept such linter? If the answer is yes, I probably will contribute to that linter 😄

AlwxSin commented:
@ruanxin I completely understand that handling errors inline is largely a matter of personal preference. Personally, I find it harder to read code where errors are inlined, because it shifts the focus from the function call itself to the error handling logic. That’s the main reason why I created this linter.

I also intentionally decided not to enable this linter by default, knowing that in existing codebases this could cause significant issues. In my view, this is a highly opinionated linter, so the decision to enable it should be left entirely to the user.

P.S.I respect the guidelines from Uber and Google, though I see them as recommendations, not mandatory rules, and I prefer a different approach in this case.

AlwxSin commented:
> ok, so we agree that if a new linter is added that handles the whole topic of inlined `if`, your linter will be deprecated inside golangci-lint and removed.
> 
> If we agree on that, we can accept this linter.

@ruanxin ldez’s comment actually touches on exactly this point — if there is a linter that comprehensively covers all cases of inline if usage and allows enforcing or forbidding any variation, then my noinlineerr linter would become unnecessary.
